src/hdmf/data_utils.py

```diff
@@ -995,6 +995,8 @@ def __deepcopy__(self, memo):
 
     def __len__(self):
         """Number of values in self.data"""
+        if self.__shape is not None:
+            return self.__shape[0]
```

Comment:
Is it possible for `self.__shape` to become out of sync with `self.data`? For example, if the user creates a DataIO without data and then sets data later but sets it to a shape not equal to self.__shape (or a dtype not equal to self.__dtype)?

Reply:
If I understand L950 correctly, no that's not possible.

Reply:
And I just added a test to make sure that doesn't happen.

Reply:
Ah, I see. Looks good. Thanks!